### PR TITLE
Added Furnace Hooks

### PR DESCRIPTION
**Pull request by CrafterKina**
I'm Japanese,so I cannot write English well.
Added ItemSmeltEvent
-Use for example in implementation of bucket of behavior when smelting the sponge.
-Pre is able to pause furnace.
-Post is able to see the result and change processing of bucket.
Added SlotFurnaceFuel Hook
-Can be increased to items to be placed in slot.(like a empty-bucket)
solve #1684
I'm sorry that I had a little disturb the Issue in mistake.

### CODE REVIEW

**Comment by OvermindDL1** [comment]
It would be nice to be able to modify the cookTime and totalCookTime from the event as well, and have the furnace use the new values.

**Comment by CrafterKina** [comment]
Why you would like to modify?
I cannot conceive what to use.

**Comment by Parker8283** [comment]
Doesn't FML have an event like this?

**Comment by CrafterKina** [comment]
Is it ”ItemSmeltedEvent”？
It is fire when the player taking out the item from results slot.
Not same.

**Comment by Parker8283** [comment]
Ah, well, the more you know :)

**Comment by Abastro** [comment]
You should squash your commits.

File: patches/minecraft/net/minecraft/inventory/SlotFurnaceFuel.java.patch [hunk 1]
```diff
@@ -0,0 +1,10 @@
+--- ../src-base/minecraft/net/minecraft/inventory/SlotFurnaceFuel.java
++++ ../src-work/minecraft/net/minecraft/inventory/SlotFurnaceFuel.java
+@@ -25,6 +25,6 @@
+ 
+     public static boolean func_178173_c_(ItemStack p_178173_0_)
+     {
+-        return p_178173_0_ != null && p_178173_0_.func_77973_b() != null && p_178173_0_.func_77973_b() == Items.field_151133_ar;
++        return net.minecraftforge.common.ForgeHooks.isBucket(p_178173_0_);
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/inventory/SlotFurnaceFuel.java.patch]: If you're gunna expand this should be any container.

File: patches/minecraft/net/minecraft/tileentity/TileEntityFurnace.java.patch [hunk 2]
```diff
@@ -10,6 +10,15 @@
...
                          }
                      }
+@@ -242,7 +241,7 @@
+ 
+                 if (this.func_145950_i() && this.func_145948_k())
+                 {
+-                    ++this.field_174906_k;
++                    if(net.minecraftforge.event.ForgeEventFactory.onItemAttemptSmelt(field_145957_n, this.field_174906_k, this.field_174905_l))++this.field_174906_k;
+ 
+                     if (this.field_174906_k == this.field_174905_l)
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/tileentity/TileEntityFurnace.java.patch]: This is fired every tick.. this seems a bit excessive. It should be fired once and left to run it's cycle.

File: patches/minecraft/net/minecraft/tileentity/TileEntityFurnace.java.patch [hunk 3]
```diff
@@ -32,3 +41,20 @@
              }
  
              if (this.field_145957_n[0].func_77973_b() == Item.func_150898_a(Blocks.field_150360_v) && this.field_145957_n[0].func_77960_j() == 1 && this.field_145957_n[1] != null && this.field_145957_n[1].func_77973_b() == Items.field_151133_ar)
+             {
+                 this.field_145957_n[1] = new ItemStack(Items.field_151131_as);
+             }
++            
++            net.minecraftforge.event.ForgeEventFactory.onItemSmelted(field_145957_n);
+
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/tileentity/TileEntityFurnace.java.patch]: If you hook above you should hook here for containers.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 4]
```diff
@@ -113,6 +114,19 @@ public static ItemStack getGrassSeed(Random rand)
         return entry.seed.copy();
     }
 
+    static final Set<ItemStack> nonFuelContainers = new HashSet();
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: We have the FulidContainer. Registry already, use it.

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 5]
```diff
@@ -413,4 +414,16 @@ public static void onPotionBrewed(ItemStack[] brewingItemStacks)
...
+    
+    public static boolean onItemAttemptSmelt(ItemStack[] smeltingItemStacks, int cookTime, int totalCookTime)
+    {
+        ItemSmeltEvent.Pre event = new ItemSmeltEvent.Pre(smeltingItemStacks, cookTime, totalCookTime);
+        MinecraftForge.EVENT_BUS.post(event);
+        return event.getResult() != Event.Result.DENY;
+    }
+    
+    public static void onItemSmelted(ItemStack[] smeltingItemStacks)
+    {
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Use your whitepsace u.u

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 6]
```diff
@@ -413,4 +414,21 @@ public static void onPotionBrewed(ItemStack[] brewingItemStacks)
...
+    public static int[] onItemAttemptSmelt(ItemStack[] smeltingItemStacks, int cookTime, int totalCookTime)
+    {
+        int[] returnTimes = new int[2];
+        ItemSmeltEvent.Pre event = new ItemSmeltEvent.Pre(smeltingItemStacks, cookTime, totalCookTime);
+        boolean cancel = MinecraftForge.EVENT_BUS.post(event);
+        returnTimes[0] = event.getCookTime();
+        returnTimes[1] = event.getTotalCookTime();
+        if(cookTime != event.getCookTime() || totalCookTime != event.getTotalCookTime())return returnTimes;
+        if(!cancel)++returnTimes[0];
+        return returnTimes;
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Use your whitepsace.

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 7]
```diff
@@ -413,4 +414,21 @@ public static void onPotionBrewed(ItemStack[] brewingItemStacks)
...
+    }
+    
+    public static void onItemSmelted(ItemStack[] smeltingItemStacks)
+    {
+        MinecraftForge.EVENT_BUS.post(new ItemSmeltEvent.Post(smeltingItemStacks));
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: More information should be exposed, namely the world location information.

**Comment by LexManos** [comment]
Lastly: Squash commits and sample mod.

**Comment by CrafterKina** [comment]
Because maybe I may not be able to accurately interpret what you are saying you, was a committed prior to squash.
Question,where should I make the SampleMod?(If possible, tell me what content is need.)
If there is no problem, I'll squash and commit.

**Comment by LexManos** [comment]
Along side the other sample mods in the test folder and it should contain a test/use of what new thing you're adding.

**Comment by CrafterKina** [comment]
Squashed.
It's OK?

**Comment by CrafterKina** [comment]
Maybe I mistook place to put the sample.

**Pull request closed without merging**